File: clp/CLP.java
import com.sun.jna.Library;
import java.nio.ByteBuffer;
import com.sun.jna.Pointer;
import java.nio.IntBuffer;
import com.sun.jna.NativeLibrary;
import com.sun.jna.ptr.IntByReference;
import com.sun.jna.Native;

public interface CLP extends Library {
	/**
	 * Original signature : <code>int clp_init()</code><br>
	 */
	int clp_init();
	/**
	 * Original signature : <code>char* clp_ver()</code><br>
	 */
	String clp_ver();
	/**
	 * Original signature : <code>int clp_pos(int)</code><br>
	 */
	int clp_pos(int id);
	/**
	 * Original signature : <code>void clp_label(int, char*)</code><br>
	 */
	void clp_label(int id, ByteBuffer out);
	/**
	 * Original signature : <code>void clp_bform(int, char*)</code><br>
	 */
	void clp_bform(int id, ByteBuffer out);
	/**
	 * Original signature : <code>void clp_forms(int, char*)</code><br>
	 */
	void clp_forms(int id, ByteBuffer out);
	/**
	 * Original signature : <code>void clp_formv(int, char*)</code><br>
	 */
	void clp_formv(int id, ByteBuffer out);
	/**
	 * Original signature : <code>void clp_vec(int, const char*, int*, int*)</code><br>
	 */
	void clp_vec(int id, java.lang.String inp, IntBuffer out, IntBuffer num);
	/**
	 * Original signature : <code>void clp_rec(const char*, int*, int*)</code><br>
	 */
	void clp_rec(java.lang.String inp, IntBuffer out, IntBuffer num);
	/**
	 * Original signature : <code>void clp_tag1(const char*, char*)</code><br>
	 */
	void clp_tag1(java.lang.String inp, ByteBuffer out);
}

File: clp.rb
require 'polish'
require 'java'
require 'jna.jar'
require 'clp/clp.jar'
require 'iconv'

java_import "com.sun.jna.Native"
java_import "CLP"
java_import "java.nio.IntBuffer"
java_import "java.nio.ByteBuffer"
java_import "java.lang.String" do
  "JString"
end

class String
  def to_bytes
    (self.split(//).map{|x| x[0]} + [0]).to_java(:byte)
  end
end

module ClpWrapper
  CLPInstance = Native.load_library("/usr/local/clp/lib/libclp_2.2.so", CLP.java_class)
  CLPInstance.clp_init
  CLPHits = {}
  CLPMisses = {}
  LabelHits = {}
  PositionHits = {}

  def self.version
    CLPInstance.clp_ver
  end

  def self.index(word)
    word = preprocess(word)
    if CLPHits[word]
      return CLPHits[word]
    elsif CLPMisses[word]
      return nil
    else
      outputs = IntBuffer.allocate(10)
      output_no = IntBuffer.allocate(1)
      CLPInstance.clp_rec(word.to_bytes, outputs, output_no)
      if output_no.get == 0
        CLPMisses[word] = true
        return nil
      else
        return CLPHits[word] = outputs.get
      end
    end
  end

  def self.flex_label(word)
    word = preprocess(word)
    return LabelHits[word] if LabelHits[word]
    index = ClpWrapper.index(word)
    return nil unless index
    outputs = ByteBuffer.allocate(10)
    CLPInstance.clp_label(index, outputs)
    result = JString.new(outputs.array).to_s[/[A-Z]+/]
    LabelHits[word] = result
  end

  def self.flex_position(word)
    word = preprocess(word)
    return PositionHits[word] if PositionHits[word]
    outputs = IntBuffer.allocate(10)
    output_no = IntBuffer.allocate(1)
    index = ClpWrapper.index(word)
    return nil unless index
    CLPInstance.clp_vec(index, word.to_bytes, outputs, output_no)
    if output_no.get == 0
      return nil
    else
      return PostionHits[word] = outputs.get
    end
  end

  def self.base_form(word)
    word = preprocess(word)
    outputs = ByteBuffer.allocate(50)
    index = ClpWrapper.index(word)
    return nil unless index
    CLPInstance.clp_bform(index, outputs)
    Iconv.conv("iso-8859-2", "utf-8", JString.new(outputs.array, "iso-8859-2").to_s.gsub("\000", ""))
  end

  def self.preprocess(word)
    word.gsub(/[., :;-?!]/, "")
  end
end

